File: algorithms/code/leetcode/lc218_the_skyline_problem/lc218_the_skyline_problem.py
#!/usr/bin/env python3
# https://leetcode.com/problems/the-skyline-problem/

import heapq
import unittest
from typing import Dict, List


class Solution:
    class Building:
        def __init__(self, left, right, height):
            self.left = left
            self.right = right
            self.height = height

        def __lt__(self, other):
            return self.height > other.height

    # pylint: disable=R0914
    def getSkyline(self, buildings: List[List[int]]) -> List[List[int]]:
        skyline: List[List[int]] = []
        if not buildings:
            return skyline
        # Build list of spots
        spots: Dict[int, List[Solution.Building]] = {}
        for building in buildings:
            if building[0] == building[1]:
                continue
            entry = Solution.Building(*building)
            for spot in building[0:2]:
                if spot not in spots:
                    spots[spot] = []
                spots[spot].append(entry)
        sorted_spots = sorted(spots.keys())
        # Prepare view
        ground = Solution.Building(0, sorted_spots[-1], 0)
        view: List[Solution.Building] = []
        view.append(ground)
        # Check all spots and build skyline
        current = 0
        for at in sorted_spots:
            for building_obj in spots[at]:
                if at == building_obj.left:
                    # Building entering the view
                    heapq.heappush(view, building_obj)
                else:
                    # Building leaving the view
                    index = view.index(building_obj)
                    view[index] = view[-1]
                    view.pop()
                    heapq.heapify(view)
            following = view[0].height
            if current != following:
                point = [at, following]
                skyline.append(point)
            current = following
        return skyline


class TestCode(unittest.TestCase):
    def generic(self, buildings, expected):
        skyline = Solution().getSkyline(buildings)
        self.assertEqual(len(skyline), len(expected))
        for i, expected_i in enumerate(expected):
            self.assertEqual(len(skyline[i]), len(expected_i))
            for j, expected_i_j in enumerate(expected_i):
                self.assertEqual(skyline[i][j], expected_i_j)

    def test_example_1(self):
        buildings = [
            [2, 9, 10],
            [3, 7, 15],
            [5, 12, 12],
            [15, 20, 10],
            [19, 24, 8],
        ]
        expected = [
            [2, 10],
            [3, 15],
            [7, 12],
            [12, 0],
            [15, 10],
            [20, 8],
            [24, 0],
        ]
        self.generic(buildings, expected)

    def test_example_2(self):
        buildings = [[0, 2, 3], [2, 5, 3]]
        expected = [[0, 3], [5, 0]]
        self.generic(buildings, expected)

    def test_coverage_gaps(self):
        buildings = [[0, 2, 3], [2, 5, 3], [0, 0, 10]]
        expected = [[0, 3], [5, 0]]
        self.generic(buildings, expected)
        self.assertEqual(len(Solution().getSkyline([])), 0)

File: algorithms/code/leetcode/lc218_the_skyline_problem/lc218_the_skyline_problem.rs
// https://leetcode.com/problems/the-skyline-problem/
// #hard

use std::cmp::Ordering;
use std::collections::{BinaryHeap, HashMap};

#[derive(Clone, Debug)]
struct Building {
    left: i32,
    right: i32,
    height: i32,
}

impl PartialEq for Building {
    fn eq(&self, other: &Self) -> bool {
        self.left == other.left && self.right == other.right && self.height == other.height
        // self.height == other.height
    }
}

impl Eq for Building {}

impl PartialOrd for Building {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for Building {
    fn cmp(&self, other: &Self) -> Ordering {
        self.height.cmp(&other.height)
        // other.height.cmp(&self.height)
    }
}

pub struct Solution;

impl Solution {
    pub fn get_skyline(buildings: Vec<Vec<i32>>) -> Vec<(i32, i32)> {
        let mut skyline = Vec::new();
        if buildings.is_empty() {
            return skyline;
        }
        // Build list of spots
        let mut spots: HashMap<i32, Vec<Building>> = HashMap::new();
        for building in buildings {
            if building[0] == building[1] {
                continue;
            }
            for spot in building[..2].iter() {
                if !spots.contains_key(spot) {
                    spots.insert(*spot, vec![]);
                }
                spots.get_mut(spot).unwrap().push(Building {
                    left: building[0],
                    right: building[1],
                    height: building[2],
                });
            }
        }
        let mut sorted_spots: Vec<i32> = spots.keys().cloned().collect();
        sorted_spots.sort();
        for spot in &sorted_spots {
            println!("{}: {:?}", spot, spots.get(spot));
        }
        println!("Sorted spots: {:?}", sorted_spots);
        // Prepare view
        let ground = Building {
            left: 0,
            right: *sorted_spots.last().unwrap(),
            height: 0,
        };
        let mut view: BinaryHeap<Building> = BinaryHeap::new();
        view.push(ground);
        // Check all spots and build skyline
        let mut current = 0;
        for at in sorted_spots {
            println!("At: {}", at);
            for building in spots.get(&at).unwrap() {
                if at == building.left {
                    // Building entering the view
                    view.push(building.clone());
                } else {
                    // Building leaving the view
                    let mut temporary_buildings: Vec<Building> = Vec::new();
                    while let Some(temporary_building) = view.pop() {
                        if temporary_building != *building {
                            temporary_buildings.push(temporary_building);
                        }
                    }
                    for temporary_building in temporary_buildings {
                        view.push(temporary_building);
                    }
                }
            }
            let following = view.peek().unwrap().height;
            if current != following {
                let point = (at, following);
                skyline.push(point);
            }
            current = following;
        }
        skyline
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_example_1() {
        let buildings = vec![
            vec![2, 9, 10],
            vec![3, 7, 15],
            vec![5, 12, 12],
            vec![15, 20, 10],
            vec![19, 24, 8],
        ];
        let expected = vec![
            (2, 10),
            (3, 15),
            (7, 12),
            (12, 0),
            (15, 10),
            (20, 8),
            (24, 0),
        ];
        assert_eq!(Solution::get_skyline(buildings), expected);
    }

    #[test]
    fn test_example_2() {
        let buildings = vec![vec![0, 2, 3], vec![2, 5, 3]];
        let expected = vec![(0, 3), (5, 0)];
        assert_eq!(Solution::get_skyline(buildings), expected);
    }

    #[test]
    fn test_coverage_gaps() {
        let buildings = vec![vec![2, 5, 3], vec![0, 2, 3], vec![0, 0, 10]];
        let expected = vec![(0, 3), (5, 0)];
        assert_eq!(Solution::get_skyline(buildings), expected);
        assert_eq!(Solution::get_skyline(vec![]).len(), 0);
    }
}
